### PR TITLE
Fix implicit type use analyzer.

### PR DESCRIPTION
**Pull request by wachulski**
Fixes edge cases where variable identifier name is reused in the initializer as first part of member access qualification.
Example: SomeEnum SomeEnum = SomeEnum.EnumVal1

Fixes dotnet/roslyn#26894

### Bugs this fixes

https://github.com/dotnet/roslyn/issues/26894

### Workarounds, if any

no

### Risk

?

### Performance impact

Low

### Is this a regression from a previous update?

?

### Root cause analysis

?

### How was the bug found?

?

### Test documentation updated?

n/a

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs [hunk 1]
```diff
@@ -259,9 +259,31 @@ protected override bool ShouldAnalyzeForEachStatement(ForEachStatementSyntax for
             }
 
             // variables declared using var cannot be used further in the same initialization expression.
-            if (initializer.DescendantNodesAndSelf()
-                    .Where(n => (n as IdentifierNameSyntax)?.Identifier.ValueText.Equals(identifier.ValueText) == true)
-                    .Any(n => semanticModel.GetSymbolInfo(n, cancellationToken).Symbol?.IsKind(SymbolKind.Local) == true))
+            bool isReusedInInitializer(SyntaxNode n)
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: 💡 Naming conventions are coming soon for this, but local functions should be PascalCase.

File: src/EditorFeatures/CSharpTest/Diagnostics/UseImplicitOrExplicitType/UseImplicitTypeTests.cs [hunk 2]
```diff
@@ -356,6 +357,111 @@ void M()
...
+            await TestMissingInRegularAndScriptAsync(
+@"using System;
+
+enum A { X, Y }
+
+class C
+{
+    void M()
+    {
+        [|A|] A = A.X;
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/UseImplicitOrExplicitType/UseImplicitTypeTests.cs]: :bulb: Would like to see a test for an expression of this form:

```txt
[|A|] A = X.A;
```

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs [hunk 3]
```diff
@@ -259,9 +259,31 @@ protected override bool ShouldAnalyzeForEachStatement(ForEachStatementSyntax for
...
+                    && id.Identifier.ValueText.Equals(identifier.ValueText)))
+                {
+                    return false;
+                }
+
+                // case of variable direct use: int x = x * 2;
+                if (semanticModel.GetSymbolInfo(n, cancellationToken).Symbol?.IsKind(SymbolKind.Local) == true)
+                {
+                    return true;
+                }
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: 💡 Include a blank line after the closing `}` here and at the end of the local function.

**Comment by sharwell** [comment]
Welcome to the project @wachulski 👍

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: Please don't plae nested function in-situ with the rest of the code.  It makes it very hard to read and grok the outer code flow.  If you want to use a nested function, please place after hte last 'return' in the method.

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs [hunk 4]
```diff
@@ -259,9 +259,31 @@ protected override bool ShouldAnalyzeForEachStatement(ForEachStatementSyntax for
...
 
             // variables declared using var cannot be used further in the same initialization expression.
-            if (initializer.DescendantNodesAndSelf()
-                    .Where(n => (n as IdentifierNameSyntax)?.Identifier.ValueText.Equals(identifier.ValueText) == true)
-                    .Any(n => semanticModel.GetSymbolInfo(n, cancellationToken).Symbol?.IsKind(SymbolKind.Local) == true))
+            bool isReusedInInitializer(SyntaxNode n)
+            {
+                // naming collision prerequisite
+                if (!(n is IdentifierNameSyntax id
+                    && id.Identifier.ValueText.Equals(identifier.ValueText)))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: IDE style is to put && on previous line.

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs [hunk 5]
```diff
@@ -259,9 +259,31 @@ protected override bool ShouldAnalyzeForEachStatement(ForEachStatementSyntax for
...
+                {
+                    return false;
+                }
+
+                // case of variable direct use: int x = x * 2;
+                if (semanticModel.GetSymbolInfo(n, cancellationToken).Symbol?.IsKind(SymbolKind.Local) == true)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: IsKind is safe to call on something null.  you don't need the ?.

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs [hunk 6]
```diff
@@ -259,9 +259,31 @@ protected override bool ShouldAnalyzeForEachStatement(ForEachStatementSyntax for
...
+                }
+                // case of qualification starting with the variable name: SomeEnum SomeEnum = SomeEnum.EnumVal1;
+                // note that: SomeEnum SomeEnum = global::SomeEnum.EnumVal1; // is ok and 'var' can be offered
+                // https://github.com/dotnet/roslyn/issues/26894
+                if (n.Parent is MemberAccessExpressionSyntax memberAccessParent && memberAccessParent.Expression == n)
+                {
+                    return true;
+                }
+
+                return false;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: i'm finding all the 'return trues' interspersed with 'return falses' to be confusing.  IMO, this wuld be clearer if one of these forms only occurred once at the end.

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs [hunk 7]
```diff
@@ -259,9 +259,31 @@ protected override bool ShouldAnalyzeForEachStatement(ForEachStatementSyntax for
...
+            }
+            if (initializer.DescendantNodesAndSelf().Any(isReusedInInitializer))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: newline after }

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs (hunk 4 above)
Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: I would prefer this not change. No need to amend previous commit for this.

Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: > if we don't change it, then i would ask that we please indent things appropriately.

The indentation is correct at 4 spaces.

Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: > I had to stare at this a lot ...

The solution I thought of was this:

```csharp
if (!(n is IdentifierNameSyntax id)
    || !id.Identifier.ValueText.Equals(identifier.ValueText))
{
}
```

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: > The indentation is always correct at 4 spaces (the only real exception in C# code is LINQ queries). Others are occasionally acceptable as well, but 4 spaces is never wrong.

Indentation of subexpressions within parenthesized expressions should align.  Otherwise you end up with code that is **super** confusing.  Is the subsequent boolean clause a sibling of the first preceding clause?  Or is it a sibling of a nested clause?  i.e.  when i see:

```c#
if (!(a...
    && b
```

my brain immediately starts pulling this apart as   ```!(a...```  as one clause and ```b``` as hte other.

But that's not hte case here.  ```&& b``` is associated with ```a``` and the ```!``` affects all of it. 

This is one of hte reasons i often push on people to not create overly complex boolean expressions as they end up hard to parse and hard to maintain over time.  To me, i would prefer something like:

```c#
if (!(n is ...
      id...)
```

Or, alternatively, just do:

```c#
var identifierNameMatches = n is IdentifierNameSyntax id && id.Identifier.ValueText.Equals(identifier.ValueText);
if (!identifierNameMatches)
{
...
```

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: > The indentation is always correct at 4 spaces

Note: i disagree strongly with this.   And we've had enough bugs and confusion in Roslyn itself with code that follows this that i think we can apply sensible rules to complex expressions to aid in readability and maintainability.  

What we could say is: at a minimum indentation should be 4 spaces.  However, indentation should be applied to most bring the highest clarity to an expression. The more complex the expression, the more effort should be made to avoid potentially incorrect interpretations.

Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: @CyrusNajmabadi None of that is considered by the formatter that runs during typing in the IDE. Please feel free to file a feature request related to it, but no alignment-related item will become part of the code style checked during review prior to it being naturally supported by the IDE. Currently the only naturally-supported items in C# are:

* Alignment of LINQ clauses
* Indentation by 4 spaces relative to previous line

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: > None of that is considered by the formatter that runs during typing in the IDE.

That's not true.  The formatter will not *format* in that manner.  But it will absolutely leave code alone that *is* formatted in that manner.  This is very intentional as attempting to codify all these potential relations has been out of scope of the formatter.  While, at the same time, it's been imperative to recognize that people *will* format in certain ways for clarity and the formatter should not undo that.

> but no alignment-related item will become part of the code style checked during review prior to it being naturally supported by the IDE or by an analyzer added to this repository

This is a pretty poor position to take.  It is completely reasonable to ask for things that help clarify the code and there are patterns and practices the IDE has followed for 20+ years.  Just ignoring that all out because we don't have a a mechanized way to enforce things is throwing the baby out with teh bathwater.

>  Currently the only naturally-supported items in C# 

Yes.  And it's fine for people to just use that approach by default.  However, in cases where code clarity is improved, it's utterly fine and reasonable to ask and to update the code.  Style/formatting are not *only* the thins that we have gotten the time to codify.  They're definitely things that we've institutionally adopted and which have a recognized value.  My position would be:

As long as it isn't onerous, we should err toward keeping the style consistent in hte IDE, and creating code that is readable, until the point that we can just have this enforced.  Otherwise, we just allow code readability to degrade in the meantime in areas that people aren't investing in.  

This does mean extra vigilance in code reviews.  But that's fine given the benefits to the health of the codebase (which shoudl be one of the highest priorities to the team, and one of the main goals of a PR in the first place).

File: src/EditorFeatures/CSharpTest/Diagnostics/UseImplicitOrExplicitType/UseImplicitTypeTests.cs (hunk 2 above)
Comment (wachulski) [reply src/EditorFeatures/CSharpTest/Diagnostics/UseImplicitOrExplicitType/UseImplicitTypeTests.cs]: sure thing, great idea

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs (hunk 6 above)
Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: Works for me 😄

File: src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs]: In combination with your proposal in https://github.com/dotnet/roslyn/pull/27897#discussion_r195865680, let's inline it for now. 👍

**Comment by sharwell** [comment]
@dotnet-bot retest this please

**Comment by jinujoseph** [comment]
Thanks for the fix @wachulski
